src/hash.cr

```diff
@@ -1343,6 +1343,15 @@ class Hash(K, V)
     self
   end
 
+  # Adds the contents of *other* to this hash. 
+  # The value of each duplicate key is determined by calling the block with the key, its value in this hash and its value in *other*.
+  #
+  #```
+  # hash = { "a" => 100, "b" => 200 }
+  # other = { "b" => 254, "c" => 300 }
+  # hash.merge!(other) { |key, v1, v2| v1 }
+  # hash # => {"a" => 100, "b" => 200, "c" => 300}
+  #```
   def merge!(other : Hash, &block)
```

Comment:
You can also add documentation for the block types:
```suggestion
   def merge!(other : Hash, & : (K, V, V) -> V) : self
```

Reply:
Are you sure? Types were specified, but removed in https://github.com/crystal-lang/crystal/commit/3c17d0708a49a6d2b931b0f6a7b808110bb2f440 (#1693). And it's sad for me, i read discussion here and say wow. It's seems like a very big bug for me - someone should fix this recursive types or it must be removed from lang, if they block clear type safety like here...

Reply:
Wow, amazing that you digged up that old issue.

But anyway, even without recursive aliases, the type restriction is incorrect, because the generic arguments of `other` don't need to equal those of `self`, they can also be smaller.
For example you can merge `Hash(String, Bool)` into `Hash(String | Int32, Bool | Nil)`. This can't be expressed currently, but is tracked in #3803. So, yeah let's leave it for now.

Reply:
At least the new syntax for non-captured blocks can be used.